databroker/src/grpc/kuksa_val_v2/val.rs

```diff
@@ -53,69 +64,402 @@ impl proto::val_server::Val for broker::DataBroker {
+    ///                  (30, Datapoint),
+    ///                  (45, Datapoint),
+    ///                  ...
+    ///              }
+    ///          }
+    ///      }
+    ///
+    ///      `response`:
+    ///      OpenProviderStreamStream {
+    ///          action: PublishValuesResponse {
+    ///              request_id: 1,
+    ///              status: {
+    ///                  (If errors) {
+    ///                      (30, Error),
+    ///                      (45, Error),
+    ///                      ...
+    ///                  }
+    ///              }
+    ///          }
+    ///      }
+    ///      ```
+    ///
+    /// 3. **Receive Actuations**:
+    ///    - The provider receives actuation requests from the databroker.
+    ///
+    ///      ```
+    ///     `request`:
+    ///      OpenProviderStreamRequest {
+    ///          action: BatchActuateStreamResponse { }
+    ///      }
+    ///
+    ///      `response`:
+    ///      OpenProviderStreamStream {
+    ///          action: BatchActuateStreamRequest {
+    ///              actuate_requests: {
+    ///                  (30, Value),
+    ///                  (45, Value),
+    ///                  ...
+    ///              }
+    ///          }
+    ///      }
+    ///      ```
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - The request should contain the necessary permissions if the databroker is started with secure mode.
+    ///   The request `OpenProviderStreamRequest` can contain messages according to the action:
+    ///   - Action 1: `ProvidedActuation`
+    ///   - Action 2: `PublishValuesRequest`
+    ///   - Action 3: `BatchActuateStreamResponse`
+    ///
+    /// # Errors
+    ///
+    /// The open stream is used for request / response type communication between the
+    /// provider and server (where the initiator of a request can vary).
+    /// Errors are communicated as messages in the stream.
     async fn open_provider_stream(
         &self,
         request: tonic::Request<tonic::Streaming<proto::OpenProviderStreamRequest>>,
-    ) -> Result<tonic::Response<Self::OpenProviderStreamStream>, tonic::Status>
-    {
-        Err(tonic::Status::new(Code::Unimplemented, "Unimplemented"))
+    ) -> Result<tonic::Response<Self::OpenProviderStreamStream>, tonic::Status> {
+        debug!(?request);
+        let permissions = match request.extensions().get::<Permissions>() {
+            Some(permissions) => {
+                debug!(?permissions);
+                permissions.clone()
+            }
+            None => return Err(tonic::Status::unauthenticated("Unauthenticated")),
+        };
+
+        // Should databroker register internally here new opened streams????
+        // The provided actuation will take ownership over the actuators but what happens
+        // if a provider is publishing sensor values and the stream is closed?
+        // How will the application know that there is no provider and should stop the subscription?
```

Comment:
thats difficult. Because only way would be to constantly check if the channel is closed or store the channels like in the subscribe function

Reply:
We dont need actually to constantly check it, with subscribe method, when the call is closed, the subscription is removed, so we would have to do it some how the same way, store the ownership sender with the signal id. But just a comment, we can discuss later.